jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java

```diff
@@ -54,22 +54,66 @@ public static String dump(Resource resource)
     /**
      * Return the Path corresponding to this resource.
      *
-     * @return the path.
+     * @return the path or null if there is no Path representation.
      */
     public abstract Path getPath();
 
     /**
      * Return true if this resource is contained in the Resource r, either because
      * r is a folder or a jar file or any form of resource capable of containing other resources.
      *
-     * @param r the containing resource
+     * @param container the containing resource
      * @return true if this Resource is contained, false otherwise
+     * @see #contains(Resource)
      */
-    public abstract boolean isContainedIn(Resource r);
+    public boolean isContainedIn(Resource container)
+    {
+        return container != null && container.contains(this);
+    }
+
+    /**
+     * Return true if this resource deeply contains the other Resource.  This resource must be
+     * a directory or a jar file or any form of resource capable of containing other resources.
+     *
+     * @param other the resource
+     * @return true if this Resource is deeply contains the other Resource, false otherwise
+     * @see #isContainedIn(Resource)
+     */
+    public boolean contains(Resource other)
+    {
+        if (other == null)
+            return false;
+
+        Path thisPath = getPath();
+        if (thisPath == null)
+            throw new UnsupportedOperationException("Resources without a Path must implement contains");
+
+        Path otherPath = other.getPath();
+        return otherPath != null && otherPath.startsWith(thisPath);
+    }
+
+    /**
+     * Get the relative path from this Resource to a possibly contained resource.
+     * @param other The other resource that may be contained in this resource
+     * @return a relative Path representing the path from this resource to the other resource.
+     */
+    public Path getPathTo(Resource other)
+    {
+        Path thisPath = getPath();
+        if (thisPath == null)
+            throw new UnsupportedOperationException("Resources without a Path must implement getPathTo");
+
+        Path otherPath = other.getPath();
+        if (otherPath == null)
+            return null;
+
+        return thisPath.relativize(otherPath);
```

Comment:
This also breaks when the two resources are in different locations.

Eg:

* Resource A is `/opt/www/jetty-base/`
* Resource B is `/home/user/tmp/index.html`

The result here is nonsense for our purposes.

Also doesn't work reliably if the the two Resources are different filesystem types.

Eg:

* Resource A is `jar:///path/to/lib.jar!/deep/`
* Resource B is `/home/user/tmp/index.html`

That will trigger an exception in the `thisPath.relativize(otherPath)`

You seem to have created this only for the AnnotationParser, seemingly because you didn't want to use the 2 argument FileId.isHidden() method?

Reply:
It is a useful method to have.    If resources are on different file systems, then a null return what we need.   I'll experiment with avoiding the exception and returning null in those situations.